### ISSUE TITLE
[3.5.RC3]Panning with <space> in tilemap will make scenes goes wrong.

### ISSUE THREAD

**Issue by gu-fan** [godotengine/godot#61840]
### Godot version

3.5.RC3 / 3.5.RC2 / 3.5.RC3.mono/3.4.4.stable.mono

### System information

Linux 64, Mint 20

### Issue description

When edit with tilemap's tile selected, pressing `<Space>` to pan around sometimes will cause the selected tile painted and like pressing `<Shift>`,
But if you press `<Ctrl-Z>` to undo, it can not undo and trigger an error:

````
 core/undo_redo.cpp:362 - Condition "action_level > 0" is true. Returned: ""
 core/undo_redo.cpp:335 - Condition "action_level > 0" is true. Returned: false
````
then this scene cannot editing tileset ( tileset bitmask will be freezed #61684)

furthermore, open new scene will make current scene became empty

All thing will be recovered after reload current project

![image](https://user-images.githubusercontent.com/579129/172773607-ff721bc6-8d25-47b2-a195-42a2e2d3a196.png)

### Steps to reproduce

1. open main.scene.
2. select tile
3. press `<Space>` to pan around and press `<Leftmouse>` to navigate on map
4. sometimes it will paint the tile like pressing shift
5. `<ctrl-z>` to undo is unavailable and the scene goes wrong

### Minimal reproduction project

[godot_bug_tilemap.zip](https://github.com/godotengine/godot/files/8867251/godot_bug_tilemap.zip)

**Comment by gu-fan** [comment]
UPDATE: 
the detailed step that can reproduce this.
1. select a tile
2. press `<leftmouse>` on map 
3. press `<Space>`
4. release `<leftmouse>`
5. move mouse to another place
6. release `<Space>`

**Comment by timothyqiu** [comment]
Closing the tilemap scene also triggers a `heap-use-after-free` on a sanitizer build.

<details><summary>Sanitizer Output</summary>

```
=================================================================
==484857==ERROR: AddressSanitizer: heap-use-after-free on address 0x617001672d58 at pc 0x000001ad8e6d bp 0x7ffe648f48e0 sp 0x7ffe648f48d0
READ of size 8 at 0x617001672d58 thread T0
    #0 0x1ad8e6c in Object::get_instance_id() const core/object.h:594
    #1 0x50f69ae in SceneTreeEditor::_compute_hash(Node*, unsigned long&) editor/scene_tree_editor.cpp:600
    #2 0x50f6d57 in SceneTreeEditor::_test_update_tree() editor/scene_tree_editor.cpp:623
    #3 0x1a79d14 in MethodBind0::call(Object*, Variant const**, int, Variant::CallError&) core/method_bind.gen.inc:59
    #4 0x8c88ff8 in Object::call(StringName const&, Variant const**, int, Variant::CallError&) core/object.cpp:918
    #5 0x8c72814 in MessageQueue::_call_function(Object*, StringName const&, Variant const*, int, bool) core/message_queue.cpp:241
    #6 0x8c72f38 in MessageQueue::flush() core/message_queue.cpp:284
    #7 0x65daf70 in SceneTree::iteration(float) scene/main/scene_tree.cpp:563
    #8 0x195630b in Main::iteration() main/main.cpp:2255
    #9 0x18c27e3 in OS_X11::run() platform/x11/os_x11.cpp:3933
    #10 0x188a1ab in main platform/x11/godot_x11.cpp:55
    #11 0x7efcd6c2928f  (/usr/lib/libc.so.6+0x2928f)
    #12 0x7efcd6c29349 in __libc_start_main (/usr/lib/libc.so.6+0x29349)
    #13 0x1889ef4 in _start (/home/timothy/repos/godot-3.x/bin/godot.x11.tools.64s+0x1889ef4)

0x617001672d58 is located 88 bytes inside of 744-byte region [0x617001672d00,0x617001672fe8)
freed by thread T0 here:
    #0 0x7efcd7abe672 in __interceptor_free /usr/src/debug/gcc/libsanitizer/asan/asan_malloc_linux.cpp:52
    #1 0x8fa5178 in Memory::free_static(void*, bool) core/os/memory.cpp:168
    #2 0x1d5c403 in void memdelete<Node>(Node*) core/os/memory.h:117
    #3 0x48a2327 in EditorData::remove_scene(int) editor/editor_data.cpp:537
    #4 0x4ae68a3 in EditorNode::_remove_edited_scene(bool) editor/editor_node.cpp:3360
    #5 0x4ae6931 in EditorNode::_remove_scene(int, bool) editor/editor_node.cpp:3372
    #6 0x4add0cb in EditorNode::_discard_changes(String const&) editor/editor_node.cpp:3008
    #7 0x4b06cc5 in EditorNode::_scene_tab_closed(int, int) editor/editor_node.cpp:4914
    #8 0x2034c89 in MethodBind2<int, int>::call(Object*, Variant const**, int, Variant::CallError&) core/method_bind.gen.inc:1491
    #9 0x8c88ff8 in Object::call(StringName const&, Variant const**, int, Variant::CallError&) core/object.cpp:918
    #10 0x8c8dd71 in Object::emit_signal(StringName const&, Variant const**, int) core/object.cpp:1230
    #11 0x8c8ee2b in Object::emit_signal(StringName const&, Variant const&, Variant const&, Variant const&, Variant const&, Variant const&, Variant const&, Variant const&, Variant const&) core/object.cpp:1285
    #12 0x6a8c546 in Tabs::_gui_input(Ref<InputEvent> const&) scene/gui/tabs.cpp:153
    #13 0x2040de9 in MethodBind1<Ref<InputEvent> const&>::call(Object*, Variant const**, int, Variant::CallError&) core/method_bind.gen.inc:759
    #14 0x8c86e26 in Object::call_multilevel(StringName const&, Variant const**, int) core/object.cpp:766
    #15 0x8c88638 in Object::call_multilevel(StringName const&, Variant const&, Variant const&, Variant const&, Variant const&, Variant const&, Variant const&, Variant const&, Variant const&) core/object.cpp:863
    #16 0x664dbe5 in Viewport::_gui_call_input(Control*, Ref<InputEvent> const&) scene/main/viewport.cpp:1659
    #17 0x665220b in Viewport::_gui_input_event(Ref<InputEvent>) scene/main/viewport.cpp:2047
    #18 0x665efb3 in Viewport::input(Ref<InputEvent> const&) scene/main/viewport.cpp:2867
    #19 0x6649f66 in Viewport::_vp_input(Ref<InputEvent> const&) scene/main/viewport.cpp:1425
    #20 0x2040de9 in MethodBind1<Ref<InputEvent> const&>::call(Object*, Variant const**, int, Variant::CallError&) core/method_bind.gen.inc:759
    #21 0x8c88ff8 in Object::call(StringName const&, Variant const**, int, Variant::CallError&) core/object.cpp:918
    #22 0x8c8826c in Object::call(StringName const&, Variant const&, Variant const&, Variant const&, Variant const&, Variant const&, Variant const&, Variant const&, Variant const&) core/object.cpp:847
    #23 0x65d724a in SceneTree::call_group_flags(unsigned int, StringName const&, StringName const&, Variant const&, Variant const&, Variant const&, Variant const&, Variant const&, Variant const&, Variant const&, Variant const&) scene/main/scene_tree.cpp:304
    #24 0x65d90e0 in SceneTree::input_event(Ref<InputEvent> const&) scene/main/scene_tree.cpp:463
    #25 0x18ffc41 in InputDefault::_parse_input_event_impl(Ref<InputEvent> const&, bool) main/input_default.cpp:491
    #26 0x19040be in InputDefault::flush_buffered_events() main/input_default.cpp:713
    #27 0x18b67d3 in OS_X11::process_xevents() platform/x11/os_x11.cpp:3074
    #28 0x18c27a4 in OS_X11::run() platform/x11/os_x11.cpp:3929
    #29 0x188a1ab in main platform/x11/godot_x11.cpp:55

previously allocated by thread T0 here:
    #0 0x7efcd7abfa89 in __interceptor_malloc /usr/src/debug/gcc/libsanitizer/asan/asan_malloc_linux.cpp:69
    #1 0x8fa45ba in Memory::alloc_static(unsigned long, bool) core/os/memory.cpp:75
    #2 0x8fa44d8 in operator new(unsigned long, char const*) core/os/memory.cpp:40
    #3 0x650f423 in Object* ClassDB::creator<Node2D>() core/class_db.h:137
    #4 0x8b4c09e in ClassDB::instance(StringName const&) core/class_db.cpp:520
    #5 0x7a3d3fe in SceneState::instance(SceneState::GenEditState) const scene/resources/packed_scene.cpp:182
    #6 0x7a556c6 in PackedScene::instance(PackedScene::GenEditState) const scene/resources/packed_scene.cpp:1598
    #7 0x4aeba14 in EditorNode::load_scene(String const&, bool, bool, bool, bool, bool) editor/editor_node.cpp:3663
    #8 0x4aa8761 in EditorNode::_sources_changed(bool) editor/editor_node.cpp:862
    #9 0x1df93bf in MethodBind1<bool>::call(Object*, Variant const**, int, Variant::CallError&) core/method_bind.gen.inc:759
    #10 0x8c88ff8 in Object::call(StringName const&, Variant const**, int, Variant::CallError&) core/object.cpp:918
    #11 0x8c8dd71 in Object::emit_signal(StringName const&, Variant const**, int) core/object.cpp:1230
    #12 0x8c8ee2b in Object::emit_signal(StringName const&, Variant const&, Variant const&, Variant const&, Variant const&, Variant const&, Variant const&, Variant const&, Variant const&) core/object.cpp:1285
    #13 0x49a331a in EditorFileSystem::_notification(int) editor/editor_file_system.cpp:1128
    #14 0x49c2ba9 in EditorFileSystem::_notificationv(int, bool) (/home/timothy/repos/godot-3.x/bin/godot.x11.tools.64s+0x49c2ba9)
    #15 0x8c892ad in Object::notification(int, bool) core/object.cpp:927
    #16 0x65e1dcd in SceneTree::_notify_group_pause(StringName const&, int) scene/main/scene_tree.cpp:1105
    #17 0x65dc3d3 in SceneTree::idle(float) scene/main/scene_tree.cpp:608
    #18 0x19569f9 in Main::iteration() main/main.cpp:2286
    #19 0x18c27e3 in OS_X11::run() platform/x11/os_x11.cpp:3933
    #20 0x188a1ab in main platform/x11/godot_x11.cpp:55
    #21 0x7efcd6c2928f  (/usr/lib/libc.so.6+0x2928f)

SUMMARY: AddressSanitizer: heap-use-after-free core/object.h:594 in Object::get_instance_id() const
Shadow bytes around the buggy address:
  0x0c2e802c6550: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c2e802c6560: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c2e802c6570: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c2e802c6580: fd fd fd fd fd fd fd fa fa fa fa fa fa fa fa fa
  0x0c2e802c6590: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
=>0x0c2e802c65a0: fd fd fd fd fd fd fd fd fd fd fd[fd]fd fd fd fd
  0x0c2e802c65b0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c2e802c65c0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c2e802c65d0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c2e802c65e0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c2e802c65f0: fd fd fd fd fd fd fd fd fd fd fd fd fd fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==484857==ABORTING
```

</details>

**Comment by KoBeWi** [comment]
> There seems to be a generic issue: In the 2D editor, you can box select nodes in the canvas by dragging. If you press down the left mouse button, then hold SPACE, then release your mouse button, and then release SPACE, you can see the editor still performing box selection.

This was fixed on master with #57000

**Comment by akien-mga** [comment]
Fixed by #61847.
